File: app/src/main/java/com/streamarc/tv/data/BufferLevel.kt
package com.streamarc.tv.data

/**
 * How much of a live stream the player keeps ahead. Bigger buffers ride out network hiccups and
 * let you pause for longer, at the cost of a longer start and more RAM. The byte figure is a
 * ceiling; the player also caps it to half of the app's heap. Levels with [diskMinutes] > 0 keep
 * the stream on device storage instead, and only use a modest in-memory buffer in front of it.
 */
enum class BufferLevel(
    val key: String,
    val minBufferMs: Int,
    val maxBufferMs: Int,
    val playbackMs: Int,
    val rebufferMs: Int,
    val bytes: Int,
    val diskMinutes: Int = 0,
) {
    SMALL("small", 8_000, 30_000, 1_500, 3_000, 24 shl 20),
    NORMAL("normal", 20_000, 60_000, 2_500, 5_000, 64 shl 20),
    LARGE("large", 45_000, 120_000, 4_000, 8_000, 128 shl 20),
    XLARGE("xlarge", 90_000, 300_000, 5_000, 10_000, 256 shl 20),
    MAX("max", 180_000, 600_000, 6_000, 12_000, 512 shl 20),
    DISK20("disk20", 20_000, 60_000, 2_500, 5_000, 64 shl 20, diskMinutes = 20),
    DISK30("disk30", 20_000, 60_000, 2_500, 5_000, 64 shl 20, diskMinutes = 30),
    DISK60("disk60", 20_000, 60_000, 2_500, 5_000, 64 shl 20, diskMinutes = 60);

    val onDisk: Boolean get() = diskMinutes > 0

    companion object {
        /** Huge is the default: the heap cap already scales it down on small devices. */
        fun from(key: String?): BufferLevel = entries.firstOrNull { it.key == key } ?: MAX
    }
}

File: desktop/Shared/Data/BufferLevel.cs
namespace StreamArcTV.Data;

/// <summary>
/// How much of a stream the player keeps ahead. Bigger buffers ride out network hiccups and let
/// you pause for longer, at the cost of more RAM. Playback starts as soon as one second is ready
/// (<see cref="PlaybackMs"/>); the level only changes how far ahead it keeps downloading. Levels
/// with <see cref="DiskMinutes"/> &gt; 0 keep the stream on disk instead, and only use a modest
/// in-memory buffer in front of it.
/// </summary>
public sealed class BufferLevel
{
    public string Key { get; }
    public int MinBufferMs { get; }
    public int MaxBufferMs { get; }
    public int PlaybackMs { get; }
    public int RebufferMs { get; }
    public int Bytes { get; }
    public int DiskMinutes { get; }
    public string Label { get; }

    private BufferLevel(string key, int min, int max, int playback, int rebuffer, int bytes, string label, int diskMinutes = 0)
    {
        Key = key; MinBufferMs = min; MaxBufferMs = max; PlaybackMs = playback; RebufferMs = rebuffer; Bytes = bytes; Label = label; DiskMinutes = diskMinutes;
    }

    public bool OnDisk => DiskMinutes > 0;

    /// Short name shown in Settings (the part before the parenthesis).
    public string ShortLabel => Label.Split(" (")[0];

    public static readonly BufferLevel SMALL = new("small", 8_000, 30_000, 1_000, 1_000, 24 << 20, "Small (30 s ahead, fastest start)");
    public static readonly BufferLevel NORMAL = new("normal", 20_000, 60_000, 1_000, 1_000, 64 << 20, "Normal (1 min ahead)");
    public static readonly BufferLevel LARGE = new("large", 45_000, 120_000, 1_000, 1_000, 128 << 20, "Large (2 min ahead, pause up to 2 min)");
    public static readonly BufferLevel XLARGE = new("xlarge", 90_000, 300_000, 1_000, 1_000, 256 << 20, "Very large (5 min ahead, pause up to 5 min)");
    public static readonly BufferLevel MAX = new("max", 180_000, 600_000, 1_000, 1_000, 512 << 20, "Huge (10 min ahead, pause up to 10 min) – default");
    public static readonly BufferLevel DISK20 = new("disk20", 20_000, 60_000, 1_000, 1_000, 64 << 20, "20 min (written to disk)", 20);
    public static readonly BufferLevel DISK30 = new("disk30", 20_000, 60_000, 1_000, 1_000, 64 << 20, "30 min (written to disk)", 30);
    public static readonly BufferLevel DISK60 = new("disk60", 20_000, 60_000, 1_000, 1_000, 64 << 20, "1 hour (written to disk)", 60);

    public static readonly BufferLevel[] Entries = { SMALL, NORMAL, LARGE, XLARGE, MAX, DISK20, DISK30, DISK60 };

    /// Huge is the default.
    public static BufferLevel From(string? key) => Entries.FirstOrDefault(e => e.Key == key) ?? MAX;
}
